presto-spi/src/main/java/com/facebook/presto/spi/session/SessionConfigurationContext.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.session;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+
+public final class SessionConfigurationContext
+{
+    private final String user;
+    private final Optional<String> source;
+    private final Optional<String> catalog;
+    private final Optional<String> schema;
+    private final Set<String> clientTags;
+
+    public SessionConfigurationContext(String user, Optional<String> source, Optional<String> catalog, Optional<String> schema, Set<String> clientTags)
+    {
+        this.user = requireNonNull(user, "user is null");
+        this.source = requireNonNull(source, "source is null");
+        this.catalog = requireNonNull(catalog, "catalog is null");
+        this.schema = requireNonNull(schema, "schema is null");
+        this.clientTags = unmodifiableSet(requireNonNull(clientTags, "clientTags is null"));
```

Comment:
This needs to make a defensive copy in addition to wrapping in unmodifiable (the caller could change the provided set)

```java
this.clientTags = unmodifiableSet(new HashSet<>(requireNonNull(clientTags, "clientTags is null")));